rust/chains/abacus-ethereum/src/provider.rs

```diff
@@ -0,0 +1,119 @@
+use async_trait::async_trait;
+use std::fmt::Debug;
+use std::sync::Arc;
+
+use ethers::prelude::{Middleware, H256, U256};
+use eyre::eyre;
+
+use crate::MakeableWithProvider;
+use abacus_core::{
+    AbacusChain, AbacusProvider, BlockInfo, ContractLocator, TxnInfo, TxnReceiptInfo,
+};
+
+/// Connection to an ethereum provider. Useful for querying information about
+/// the blockchain.
+#[derive(Debug, Clone)]
+pub struct EthereumProvider<M>
+where
+    M: Middleware,
+{
+    provider: Arc<M>,
+    chain_name: String,
+    domain: u32,
+}
+
+impl<M> AbacusChain for EthereumProvider<M>
+where
+    M: Middleware + 'static,
+{
+    fn chain_name(&self) -> &str {
+        &self.chain_name
+    }
+
+    fn local_domain(&self) -> u32 {
+        self.domain
+    }
+}
+
+#[async_trait]
+impl<M> AbacusProvider for EthereumProvider<M>
+where
+    M: Middleware + 'static,
+{
+    async fn get_block_by_hash(&self, hash: &H256) -> eyre::Result<BlockInfo> {
+        let block = self
+            .provider
+            .get_block(*hash)
+            .await?
+            .ok_or_else(|| eyre!("Could not find block with hash {}", hash))?;
+        assert_eq!(block.hash.as_ref().unwrap(), hash);
```

Comment:
should we panic here & line 84 or just return an `Err`?

Don't feel too strongly tbh

Reply:
I guess it depends on how much we trust the providers. It probably should be a `debug_assert_eq` if we trust them, and return an error if we don't. Thoughts?

Reply:
Now that we use quorumprovider i have a higher faith that the quorum will only be reached for valid data, so let's do `debug_assert_eq` ?

Reply:
Removed